### PR TITLE
Add UniqueId to PartitionKey for some Dendrite tables (where possible)

### PR DESCRIPTION
**Pull request by alexfca**
The PartitonKey scheme currently uses the collectionName which only includes the Tenant and TableName.
For scaling best practices, the more unique the PK is the more Cosmos can scale which also improves perf.

This PR introduces a specific (UniqueId) PK scheme where possible using the following criteria
- The UniqueId does not change
- The UniqueId is in a unique key constraint
- The UniqueId is used in every SQL query

Every table was evaluated and checked whether a uniqueid can be introduced for the PartitionKey.
- When there was an obvious uniqueid it was added
- When there was no obvious uniqueid and the table was "highly used" a comment was added according to
https://criticalarc.atlassian.net/wiki/spaces/~343945362/pages/2211774465/CosmosDB+Refactor
- When there was no obvious uniqueid it was left to use CollectionName

So, for tables where a uniqueid can be included the PK looks like
matrix_naffka_messages,test.criticalarc.com,**1**
where "1" is the uniqueid
`    {
        "id": "matrix_naffka_messages,test.criticalarc.com,1_0",
        "_pk": "matrix_naffka_messages,test.criticalarc.com,1",
        "_sid": "test.criticalarc.com",`

As well as this the PerformQuery() and PerformQueryAllPartitions() include validation to ensure that the correct SQL has been used.
This was mainly done to ensure the c._sid is always present for cross-partition queries 

- Update specific PKs to be item specific where possible
- Add validation to the PerformQueryXX methods
- Fix queries that fail validation

### CODE REVIEW

File: federationsender/storage/cosmosdb/inbound_peeks_table.go [hunk 1]
```diff
@@ -92,8 +92,9 @@ func (s *inboundPeeksStatements) getCollectionName() string {
 	return cosmosdbapi.GetCollectionName(s.db.databaseName, s.tableName)
 }
 
-func (s *inboundPeeksStatements) getPartitionKey() string {
-	return cosmosdbapi.GetPartitionKeyByCollection(s.db.cosmosConfig.TenantName, s.getCollectionName())
+func (s *inboundPeeksStatements) getPartitionKey(roomId string) string {
+	uniqueId := roomId
```
Comment (alexfca) [inline federationsender/storage/cosmosdb/inbound_peeks_table.go:96]: This uses the "roomId" as the UniqueId

File: internal/cosmosdbapi/client.go [hunk 2]
```diff
@@ -55,9 +57,13 @@ func PerformQuery(ctx context.Context,
 	qryString string,
 	params map[string]interface{},
 	response interface{}) error {
+	err := validateQuery(qryString)
```
Comment (alexfca) [inline internal/cosmosdbapi/client.go:60]: Some basic validation for "normal" Queries

File: internal/cosmosdbapi/client.go [hunk 3]
```diff
@@ -74,9 +80,13 @@ func PerformQueryAllPartitions(ctx context.Context,
 	qryString string,
 	params map[string]interface{},
 	response interface{}) error {
+	err := validateQueryAllPartitions(qryString)
```
Comment (alexfca) [inline internal/cosmosdbapi/client.go:83]: Extra validation for "crioss-partition" Queries

File: internal/cosmosdbapi/client.go [hunk 4]
```diff
@@ -130,3 +140,30 @@ func GetDocumentOrNil(connection CosmosConnection, config CosmosConfig, ctx cont
...
+	}
+	return nil
+}
+
+func validateQueryAllPartitions(qryString string) error {
+	err := validateQuery(qryString)
+	if err != nil {
+		return err
+	}
+	if strings.Contains(qryString, " top ") {
```
Comment (alexfca) [inline internal/cosmosdbapi/client.go:159]: Cross Partition Queries
- cannot include top
- cannot include order by
- must include our "SZ" id "c._sid"

**Review by scottj-ca — APPROVED** [review]

**Review by jaredjn — APPROVED** [review]

File: internal/cosmosdbapi/client.go (hunk 4 above)
Comment (jahmai-ca) [reply internal/cosmosdbapi/client.go:159]: point 3 makes sense, the other 2 are they limits on the SDK or something you've decided to improve performance, we do order by with cross-partition with the C# sdk

Comment (alexfca) [reply internal/cosmosdbapi/client.go:159]: I found that when using cross-partitioned queries and including these in the SQL that I get 0 results with no error.
When I removed both of these, all the data came back, so I didnt wanna change any SQL in this rev.
I found some relevant info on the web but wasnt sure if its still relevant, but my results did seem to correlate..
https://stackoverflow.com/questions/50240232/cosmos-db-rest-api-order-by-with-partitioning

Happy to revisit this if required, but for now it seems like maybe the .Net SDK does some "magic" maybe?

Comment (jahmai-ca) [reply internal/cosmosdbapi/client.go:159]: Possibly, the client needs to coalesce results for cross partition ordering, so probably the SDK for this doesn't do that

**Review by jahmai-ca — APPROVED** [review]

**Pull request merged**